File: kafkaStreamsModelServer/src/main/java/com/lightbend/java/kafkastreams/modelserver/memorystore/DataProcessor.java
package com.lightbend.java.kafkastreams.modelserver.memorystore;

import com.lightbend.java.model.ServingResult;
import com.lightbend.java.kafkastreams.store.StoreState;
import com.lightbend.java.model.DataConverter;
import com.lightbend.java.model.ModelServingInfo;
import com.lightbend.model.Winerecord;
import org.apache.kafka.streams.processor.AbstractProcessor;
import org.apache.kafka.streams.processor.ProcessorContext;

import java.util.Optional;

/**
 * Created by boris on 7/12/17.
 * used
 * https://github.com/bbejeck/kafka-streams/blob/master/src/main/java/bbejeck/processor/stocks/StockSummaryProcessor.java
 */
public class DataProcessor extends AbstractProcessor<byte[], byte[]> {

    private StoreState modelStore;
    private ProcessorContext ctx;

    @Override
    public void process(byte[] key, byte[] value) {
        Optional<Winerecord.WineRecord> dataRecord = DataConverter.convertData(value);
        if(!dataRecord.isPresent()) {
            return;                                 // Bad record
        }
        if(modelStore.getNewModel() != null){
            // update the model
            if(modelStore.getCurrentModel() != null)
                modelStore.getCurrentModel().cleanup();
            modelStore.setCurrentModel(modelStore.getNewModel());
            modelStore.setCurrentServingInfo(new ModelServingInfo(modelStore.getNewServingInfo().getName(),
                    modelStore.getNewServingInfo().getDescription(), System.currentTimeMillis()));
            modelStore.setNewServingInfo(null);
            modelStore.setNewModel(null);
        }
        // Actually score
        if(modelStore.getCurrentModel() == null) {
            // No model currently
//            System.out.println("No model available - skipping");
            ctx.forward(key, ServingResult.noModel);
        }
        else{
            // Score the model
            long start = System.currentTimeMillis();
            double quality = (double) modelStore.getCurrentModel().score(dataRecord.get());
            long duration = System.currentTimeMillis() - start;
            modelStore.getCurrentServingInfo().update(duration);
//            System.out.println("Calculated quality - " + quality + " in " + duration + "ms");
            ctx.forward(key, new ServingResult(quality, duration));
         }
         ctx.commit();
    }

    @Override
    public void init(ProcessorContext context) {
        modelStore = StoreState.getInstance();
        ctx = context;
    }
}

File: kafkaStreamsModelServer/src/main/scala/com/lightbend/scala/kafkastreams/modelserver/memorystore/DataProcessor.scala
package com.lightbend.scala.kafkastreams.modelserver.memorystore

import com.lightbend.scala.kafkastreams.store.StoreState
import com.lightbend.scala.modelServer.model.{DataRecord, ServingResult}
import org.apache.kafka.streams.processor.{AbstractProcessor, ProcessorContext}

import scala.util.Success

/**
 * The DataProcessor for the in-memory state store for the Kafka Streams example.
 * See also this example:
 * https://github.com/bbejeck/kafka-streams/blob/master/src/main/java/bbejeck/processor/stocks/StockSummaryProcessor.java
 */
class DataProcessor extends AbstractProcessor[Array[Byte], Array[Byte]]{

  private var modelStore: StoreState = null
  private var ctx: ProcessorContext = null

  // Exercise:
  // See the exercises described in com.lightbend.scala.kafkastreams.modelserver.customstore.DataProcessor.
  override def process(key: Array[Byte], value: Array[Byte]): Unit = {
    DataRecord.fromByteArray(value) match {
      case Success(dataRecord) => {
        modelStore.newModel match {
          case Some(model) => {
            // close current model first
            modelStore.currentModel match {
              case Some(m) => m.cleanup()
              case _ =>
            }
            // Update model
            modelStore.currentModel = Some(model)
            modelStore.currentState = modelStore.newState
            modelStore.newModel = None
          }
          case _ =>
        }
        modelStore.currentModel match {
          case Some(model) => {
            val start = System.currentTimeMillis()
            val quality = model.score(dataRecord).asInstanceOf[Double]
            val duration = System.currentTimeMillis() - start
//            println(s"Calculated quality - $quality calculated in $duration ms")
            modelStore.currentState = modelStore.currentState.map(_.incrementUsage(duration))
            ctx.forward(key, ServingResult(quality, duration))
          }
          case _ => {
//            println("No model available - skipping")
            ctx.forward(key, ServingResult.noModel)
          }
        }
        ctx.commit()
      }
      case _ => // error; ignore
        // Exercise:
        // Like all good production code, we're ignoring errors ;) here! That is, we filter to keep
        // messages where a `Success(x)` is returned true and ignore the `Failure(exception)` results.
        // With the topology API, this is harder to fix; what could you do in those `case _ =>` clauses??
        // See the implementation of `DataRecord`, where we inject fake errors. 
    }
  }

  override def init(context: ProcessorContext): Unit = {
    modelStore = StoreState()
    ctx = context
  }
}
